data/keystore/src/test/java/tech/pegasys/artemis/bls/keystore/KeyStoreTest.java

```diff
@@ -0,0 +1,262 @@
+    assertNotNull(keyStoreData);
+    final KdfParam param = keyStoreData.getCrypto().getKdf().getParam();
+    assertNotNull(param);
+    if (keyStoreData.getCrypto().getKdf().getCryptoFunction() == SCRYPT) {
+      assertTrue(param instanceof SCryptParam);
+    } else if (keyStoreData.getCrypto().getKdf().getCryptoFunction() == PBKDF2) {
+      assertTrue(param instanceof Pbkdf2Param);
+    } else {
+      fail("Unsupported crypto function");
+    }
+    assertTrue(keyStore.validatePassword(PASSWORD));
+    assertFalse(keyStore.validatePassword("test"));
+
+    final Bytes decryptedKey = keyStore.decrypt(PASSWORD);
+    assertEquals(BLS_PRIVATE_KEY, decryptedKey);
+  }
+
+  @ParameterizedTest
+  @MethodSource("invalidKeyStorePaths")
+  void loadingKeyStoreWithInvalidKdfParamsThrowsException(final Path invalidJsonPath) {
+    assertThrows(JsonMappingException.class, () -> KeyStoreLoader.loadFromFile(invalidJsonPath));
+  }
+
+  @ParameterizedTest
+  @MethodSource("kdfParams")
+  void encryptWithKdfAndCipherFunction(
+      final KdfParam kdfParam, final Bytes expectedChecksum, final Bytes encryptedCipherMessage) {
+    final CipherParam cipherParam = CipherParamBuilder.aCipherParam().withIv(AES_IV_PARAM).build();
+    final Cipher cipher =
+        CipherBuilder.aCipher()
+            .withCipherParam(cipherParam)
+            .withCipherFunction(AES_128_CTR)
+            .build();
+    final Crypto crypto =
+        KeyStore.encryptUsingCipherFunction(
+            BLS_PRIVATE_KEY, PASSWORD, kdfParam, cipher, ChecksumFunction.SHA256);
+
+    assertEquals(expectedChecksum, crypto.getChecksum().getMessage());
+    assertEquals(encryptedCipherMessage, crypto.getCipher().getMessage());
+  }
+
+  @ParameterizedTest
+  @MethodSource("basicKdfParam")
+  void encryptWithRandomSaltAndIv(final KdfParam kdfParam, final CipherParam cipherParam) {
+    final Cipher cipher =
+        CipherBuilder.aCipher()
+            .withCipherParam(cipherParam)
+            .withCipherFunction(AES_128_CTR)
+            .build();
+    final KeyStore encryptedKeyStore =
+        KeyStore.encrypt(BLS_PRIVATE_KEY, "test", "", kdfParam, cipher);
+    assertTrue(encryptedKeyStore.validatePassword("test"));
+    final Bytes decryptedKey = encryptedKeyStore.decrypt("test");
+    assertEquals(BLS_PRIVATE_KEY, decryptedKey);
+  }
+
+  @ParameterizedTest
+  @EnumSource(CryptoFunction.class)
+  void encryptLargeDataWithRandomSaltAndIv(final CryptoFunction cryptoFunction) {
+    final KdfParam kdfParam =
+        cryptoFunction == SCRYPT
+            ? SCryptParamBuilder.aSCryptParam().build()
+            : Pbkdf2ParamBuilder.aPbkdf2Param().build();
+    final CipherParam cipherParam = CipherParamBuilder.aCipherParam().build();
+    final Cipher cipher =
+        CipherBuilder.aCipher()
+            .withCipherParam(cipherParam)
+            .withCipherFunction(AES_128_CTR)
+            .build();
+    final Bytes largeData = Bytes.random(1024);
```

Comment:
Random in tests is a bad idea.  We should at least use an explicit seed to ensure the test is repeatable.